### PR TITLE
Vendor Generated Protobuf Code (#3947)

### LINKED ISSUE #3538: `docs.rs` cannot build `datafusion-proto` crate
**Describe the bug**

```
[INFO] [stderr]    Compiling datafusion-proto v12.0.0 (/opt/rustwide/workdir)
[INFO] [stderr] error: failed to run custom build command for `datafusion-proto v12.0.0 (/opt/rustwide/workdir)`
[INFO] [stderr] 
[INFO] [stderr] Caused by:
[INFO] [stderr]   process didn't exit successfully: `/opt/rustwide/target/debug/build/datafusion-proto-4369f54fa6f25371/build-script-build` (exit status: 1)
[INFO] [stderr]   --- stdout
[INFO] [stderr]   cargo:rerun-if-env-changed=FORCE_REBUILD
[INFO] [stderr]   cargo:rerun-if-changed=proto/datafusion.proto
[INFO] [stderr]   Running: "/usr/bin/protoc" "--include_imports" "--include_source_info" "-o" "/tmp/prost-buildQPnxTC/prost-descriptor-set" "-I" "proto" "proto/datafusion.proto"
[INFO] [stderr] 
[INFO] [stderr]   --- stderr
[INFO] [stderr]   Error: "protobuf compilation failed: Read-only file system (os error 30)"
```

**To Reproduce**

**Expected behavior**

**Additional context**

### LINKED ISSUE #3947: Don't make dependants install protoc
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

I'd like to file a PR with delta-rs (https://github.com/delta-io/delta-rs/pull/892) but it won't pass their CI because we require all our downstream dependencies to install protoc. Let's fix that.

**Describe the solution you'd like**

Have `build.rs` automatically download and extract the correct protoc distribution for the build platform into `OUT_DIR` and use that for the prost build.

**Describe alternatives you've considered**

1. require users to install it manually
2. PR prost to add this functionality
3. stop using prost

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #3538 
Closes #3947

 # Rationale for this change

We don't want to require downstreams to install PROTOC, but we also don't want to allow the generated code to get out of sync. This is the approach we use with arrow-rs, and it appears to work fairly well. Specifically we don't bundle the .proto files in the released crate, and only generate the code if the .proto files exist.

# What changes are included in this PR?

# Are there any user-facing changes?

datafusion-proto now always has a build dependency on pbjson-build. This is a fairly lightweight dependency and makes the logic significantly easier to follow.

### CODE REVIEW

**Comment by tustvold** [comment]
@avantgardnerio What do you think about this?

File: dev/release/rat_exclude_files.txt [hunk 1]
```diff
@@ -129,3 +129,5 @@ Cargo.lock
 .history
 parquet-testing/*
 *rat.txt
+datafusion/proto/src/generated/pbjson.rs
```
Comment (tustvold) [inline dev/release/rat_exclude_files.txt:132]: There is precedent in arrow for excluding generated files - https://github.com/apache/arrow-rs/blob/master/dev/release/rat_exclude_files.txt#L20

**Comment by isidentical** [comment]
@tustvold does it make sense to also include a GH action to ensure that PRs that actually touch the source also generate the latest version of the declarations? Something like the following: 
https://github.com/apache/arrow-datafusion/blob/7559c4425e6f32655c6d09e8ed17c9c51896472b/.github/workflows/rust.yml#L65-L67

**Comment by avantgardnerio** [comment]
I'd approve this PR if:

1. generated files have a formatter exception
2. there's clear doc on how to run protoc in a readme somewhere

**Comment by andygrove** [comment]
Can we provide a Dockerfile with protoc for generating the code?

**Review by avantgardnerio — APPROVED** [review]
It seems like everyone is leaning towards this solution. As stated above, I think the main concern is that dependents don't have to install protoc, and this PR achieves that goal, so I'm approving and closing mine.

**Comment by tustvold** [comment]
> generated files have a formatter exception

Already in place

> there's clear doc on how to run protoc in a readme somewhere

This is still automatically run as part of build.rs, I will add some docs though

> different versions of protoc will generate different output

Protoc is only used to generate a descriptor set, not the generated Rust code which is versioned by prost-build. The output should therefore be relatively stable.

> dependents don't have to install protoc

I suspect dependencies by git sha may still need protoc installed - as this is effectively a source dependency I think this is fine

> We should have CI prevent that from happening

I will add a CI check to here and arrow-rs

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -81,16 +81,6 @@ jobs:
       - uses: actions/checkout@v3
         with:
           submodules: true
-      - name: Install protobuf compiler
```
Comment (tustvold) [inline .github/workflows/rust.yml:84]: This doesn't appear to be necessary anymore, the version in the system package manager is recent enough (only 2 years out of date) :laughing:

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @tustvold and @avantgardnerio

**Pull request merged**